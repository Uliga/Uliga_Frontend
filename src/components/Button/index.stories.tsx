import React from "react";
import { withKnobs } from "@storybook/addon-knobs";
import Button from "./index";

export default {
  title: "components|basic/Button", // 스토리북에서 보여질 그룹과 경로를 명시
  component: Button, // 어떤 컴포넌트를 문서화 할지 명시
  decorators: [withKnobs], // 애드온 적용
};

export function Primary() {
  return <Button title="Primary" size="medium" theme="primary" />;
}
export function Secondary() {
  return <Button title="Secondary" size="medium" theme="secondary" />;
}
export function Tertiary() {
  return <Button title="Tertiary" size="medium" theme="tertiary" />;
}
export function Quaternary() {
  return <Button title="Quaternary" size="medium" theme="quaternary" />;
}
export function Unfocus() {
  return <Button title="Unfocus" size="medium" theme="unfocus" />;
}
export function Basic() {
  return <Button title="Basic" size="medium" theme="basic" />;
}
export function Disabled() {
  return <Button title="Disabled" size="medium" disabled />;
}
export function Large() {
  return <Button title="Large" size="large" />;
}
export function Medium() {
  return <Button title="Medium" size="medium" />;
}
export function Small() {
  return <Button title="Small" size="small" />;
}
export function StoryBookTest() {
  return <Button title="Primary" size="medium" theme="secondary" />;
}

export function StoryBookTest2() {
  return <Button title="Primary" size="medium" theme="secondary" />;
}
export function StoryBookTest3() {
  return <Button title="Primary" size="large" theme="secondary" />;
}
