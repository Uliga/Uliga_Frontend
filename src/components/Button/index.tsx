import React from "react";
import styled from "styled-components";
import COLORS from "../../constants/color";

type ButtonTheme =
  | "primary"
  | "secondary"
  | "tertiary"
  | "unfocus"
  | "quaternary"
  | "basic";
type ButtonSize = "large" | "medium" | "small";
type ButtonType = "button" | "submit" | "reset" | undefined;
type ButtonProps = {
  /** 버튼 안의 내용 */
  title?: string;
  /** 버튼의 테마 */
  theme?: ButtonTheme;
  /** large, medium, small 사이즈 중 택1 */
  size?: ButtonSize;
  /** 버튼 활성화 여부 */
  disabled?: boolean;
  /** 버튼 너비 */
  width?: string;
  /** 버튼 타입 */
  type?: ButtonType;
  /** 클릭했을 때 호출할 함수 */
  onClick?: (e?: React.MouseEvent<HTMLButtonElement>) => void;
  className?: string;
  dataCy?: string;
};

const Wrapper = styled.button<{
  size: ButtonSize;
  ButtonTheme: ButtonTheme;
  disabled: boolean;
  width: string;
  className?: string;
}>`
  ${({ size, ButtonTheme, width, disabled }) => `
    width: ${width};
    font-size: ${{ large: "1.2rem", medium: "1rem", small: "0.75rem" }[size]};
     padding: ${
       {
         large: "0.5rem 1.5rem",
         medium: "0.5rem 1rem",
         small: "0.5rem 0.875rem",
       }[size]
     };
     background-color: ${
       (disabled && "#F2F4F7") ||
       {
         primary: COLORS.BLUE,
         secondary: COLORS.YELLOW,
         tertiary: COLORS.WHITE,
         unfocus: COLORS.WHITE,
         quaternary: COLORS.LIGHT_BLUE,
         basic: COLORS.WHITE,
       }[ButtonTheme]
     };
     border: ${
       ButtonTheme === "tertiary" ? `0.08rem solid ${COLORS.BLUE} ` : "none"
     };
     border: ${
       ButtonTheme === "basic" ? `0.08rem solid ${COLORS.GREY[300]} ` : "none"
     };
     color: ${
       (disabled && "#CCCCCF") ||
       {
         primary: COLORS.WHITE,
         secondary: COLORS.WHITE,
         tertiary: COLORS.BLUE,
         unfocus: COLORS.BLUE,
         quaternary: COLORS.BLUE,
         basic: COLORS.GREY[400],
       }[ButtonTheme]
     };
    &:hover{
      background-color: ${
        (disabled && "#F2F4F7") ||
        {
          primary: COLORS.MEDIUM_BLUE,
          secondary: COLORS.LIGHT_YELLOW,
          tertiary: COLORS.WHITE,
          unfocus: COLORS.LIGHT_BLUE,
          quaternary: COLORS.BLUE,
          basic: COLORS.WHITE,
        }[ButtonTheme]
      };
   
     color: ${
       (disabled && "#CCCCCF") ||
       {
         primary: COLORS.WHITE,
         secondary: COLORS.WHITE,
         tertiary: COLORS.DARK_BLUE,
         unfocus: COLORS.BLUE,
         quaternary: COLORS.WHITE,
         basic: COLORS.GREY[300],
       }[ButtonTheme]
     };
     cursor: ${disabled ? "default" : "pointer"};
    }
`}
  box-sizing: border-box;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 1rem;
`;

export default function Button({
  title,
  theme = "primary",
  size = "medium",
  disabled = false,
  width,
  type = "button",
  onClick,
  className,
  dataCy = "",
}: ButtonProps) {
  return (
    <Wrapper
      className={className}
      size={size}
      ButtonTheme={theme}
      onClick={onClick}
      disabled={disabled}
      width={width!}
      type={type}
      data-cy={dataCy}
    >
      {title}
    </Wrapper>
  );
}
