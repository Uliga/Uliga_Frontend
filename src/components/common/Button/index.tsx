/* eslint-disable react/button-has-type */
import React from "react";
import styled from "styled-components";

type ButtonTheme = "primary" | "secondary" | "tertiary";
type ButtonSize = "small" | "medium" | "large";
type ButtonProps = {
  title: string;
  theme?: ButtonTheme;
  size?: ButtonSize;
  disabled?: boolean;
  width?: string;
  onClick?: (e?: React.MouseEvent<HTMLButtonElement>) => void;
};

const Wrapper = styled.button<{
  title: string;
  theme?: ButtonTheme;
  disabled?: boolean;
  width?: string;
}>`
  box-sizing: border-box;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export default function Button({
  title,
  theme = "primary",
  size,
  disabled = false,
  width,
  onClick,
}: ButtonProps) {
  return <Wrapper>{title}</Wrapper>;
}
