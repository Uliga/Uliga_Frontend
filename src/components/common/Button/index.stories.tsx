import React from "react";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import Button from "./index";

export default {
  title: "components|basic/Button", // 스토리북에서 보여질 그룹과 경로를 명시
  component: Button, // 어떤 컴포넌트를 문서화 할지 명시
  decorators: [withKnobs], // 애드온 적용
};

// // knobs
// export const Button = () => {
//   // knobs 만들기
//   const big = boolean("big", false);
//   const name = text("name", "Storybook");
//   return <Button name={name} big={big} />;
// };
// hello.story = {
//   name: "Default",
// };

export function Primary() {
  return <Button title="Button" theme="primary" />;
}
// export const Secondary = () => <Button title="Button" />;
// export const Tertiary = () => <Button title="Button" />;
// export const Unfocus = () => <Button title="Button" />;
// export const Disabled = () => <Button title="Button" />;
// export const FullWidth = () => <Button title="Button" />;
// export const CustomWidth = () => <Button title="Button" />;
