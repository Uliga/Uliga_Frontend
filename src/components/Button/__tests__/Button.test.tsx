import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Button from "../index";

// describe는 테스트 케이스를 하나로 묶어주는 단위라고 생각하면 된다.
describe("버튼 컴포넌트 단위 테스트", () => {
  // it 또는 test 둘 중 아무거나 사용해도 문제없다.
  it("버튼이 dom에 존재하는지 확인한다.", () => {
    const buttonTitle = "버튼 컴포넌트입니다.";
    // getByText 함수를 사용해서 렌더링 된 요소에서 특정 텍스트를 찾을 수 있다.
    const { getByText } = render(<Button title={buttonTitle} />);
    // 위에서 반환된 getByText함수를 사용해서 buttonTitle이라는 변수에 저장된 문자열을 가진 요소를 찾는다.
    const button = getByText(buttonTitle);
    // 해당 버튼이 dom에 존재하는지 확인한다.
    expect(button).toBeInTheDocument();
    // toMatchSnapshot Matcher를 사용해서 현재 렌더링 결과와 저장된 스냅샷을 비교하여 테스트한다.
    expect(button).toMatchSnapshot();
  });

  it("버튼이 제대로 클릭되는지 확인한다.", () => {
    const buttonTitle = "onClick event를 가지고 있는 버튼 컴포넌트입니다.";
    // jest.fn() 메서드를 사용해서 대체 콜백 함수를 생성.
    const onClickMock = jest.fn();
    const { getByText } = render(
      <Button title={buttonTitle} onClick={onClickMock} />,
    );
    const button = getByText(buttonTitle);
    // fireEvent 메서드를 이용해서(@testing-library/react의 기본 메서드) 버튼을 클릭하는 이벤트를 트리거한다.
    // 여기서 이벤트를 트리거한다는 말은 사용자가 실제로 버튼을 클릭한 것 처럼 시뮬레이션한다는 뜻이다.
    fireEvent.click(button);
    // 위에서 생성된 Mock함수가 제대로 호출되었는지 toHaveBeenCalled() 메서드로 확인한다.
    expect(onClickMock).toHaveBeenCalled();
  });

  it("버튼의 disabled 속성이 제대로 적용되었는지 확인한다.", () => {
    const buttonTitle = "disabled 속성을 가지고 있는 버튼 컴포넌트입니다.";
    const { getByText } = render(<Button title={buttonTitle} disabled />);
    const button = getByText(buttonTitle);
    // 해당 요소가 비활성화(disabled) 상태인지 Matcher를 통해 확인한다.
    expect(button).toBeDisabled();
  });

  it("버튼의 css 속성이 제대로 들어가있는지 확인한다.", () => {
    const buttonTitle =
      "secondary theme과 large size 속성을 가지고 있는 버튼 컴포넌트입니다.";
    const { getByText } = render(
      <Button title={buttonTitle} theme="secondary" size="large" />,
    );
    const button = getByText(buttonTitle);
    // toHaveStyle Matcher를 사용해서 확인한다.
    expect(button).toHaveStyle(`
      font-size: 1.2rem;
      padding: 0.5rem 1.5rem;
      background-color: #FFC144;
      color: #FFF;
    `);
    // toMatchSnapshot Matcher를 사용해서 현재 렌더링 결과와 저장된 스냅샷을 비교하여 테스트한다.
    expect(button).toMatchSnapshot();
  });
});
