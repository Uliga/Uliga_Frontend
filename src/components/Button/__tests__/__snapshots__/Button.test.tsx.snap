// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> 버튼의 css 속성이 제대로 들어가있는지 확인한다. 1`] = `
<button
  class="sc-bcXHqe DLusH"
  data-cy=""
  type="button"
>
  secondary theme과 large size 속성을 가지고 있는 버튼 컴포넌트입니다.
</button>
`;

exports[`<Button /> 버튼이 dom에 존재하는지 확인한다. 1`] = `
<button
  class="sc-bcXHqe hbqTtc"
  data-cy=""
  type="button"
>
  버튼 컴포넌트입니다.
</button>
`;
